drivers/adc/adc_mchp_xec.c

```diff
@@ -0,0 +1,334 @@
+		u32_t reserved_2:15;
+	} sar_control_reg;
+};
+
+#define ADC_XEC_REG_BASE(_dev)						\
+	((struct adc_xec_regs *)(DT_INST_0_MICROCHIP_XEC_ADC_BASE_ADDRESS))
+
+
+DEVICE_DECLARE(adc_xec);
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_xec_data *data = CONTAINER_OF(ctx, struct adc_xec_data, ctx);
+	struct adc_xec_regs *adc_regs = ADC_XEC_REG_BASE(data->dev);
+
+	data->repeat_buffer = data->buffer;
+
+	adc_regs->single_reg.single_en = data->channels;
+	adc_regs->control_reg.start_single = 1;
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_xec_data *data = CONTAINER_OF(ctx, struct adc_xec_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static int adc_xec_channel_setup(struct device *dev,
+				 const struct adc_channel_cfg *channel_cfg)
+{
+	struct adc_xec_regs *adc_regs = ADC_XEC_REG_BASE(dev);
+	u32_t ref_pos;
+	u32_t mask;
+
+	if (channel_cfg->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
+		return -EINVAL;
+	}
+
+	if (channel_cfg->channel_id >= XEC_ADC_CHANNEL_COUNT) {
+		return -EINVAL;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		return -EINVAL;
+	}
+
+	ref_pos = channel_cfg->channel_id * 2;
+	mask = ~(XEC_ADC_VREF_MASK << ref_pos);
+
+	if (channel_cfg->reference == ADC_REF_INTERNAL) {
+		adc_regs->vref_channel_reg.vref_chan &= mask;
+	} else if (channel_cfg->reference == ADC_REF_EXTERNAL0) {
+		adc_regs->vref_channel_reg.vref_chan &= mask;
+		adc_regs->vref_channel_reg.vref_chan |=
+			XEC_ADC_VREF_GPIO << ref_pos;
+	} else {
+		return -EINVAL;
+	}
+
+	adc_regs->sar_control_reg.seldiff = channel_cfg->differential;
+
+	return 0;
+}
+
+static bool adc_xec_validate_buffer_size(const struct adc_sequence *sequence)
+{
+	int chan_count = 0;
+	size_t buff_need;
+	u32_t chan_mask;
+
+	for (chan_mask = 0x80; chan_mask != 0; chan_mask >>= 1) {
+		if (chan_mask & sequence->channels) {
+			chan_count++;
+		}
+	}
+
+	buff_need = chan_count * sizeof(u16_t);
+
+	if (sequence->options) {
+		buff_need *= 1 + sequence->options->extra_samplings;
+	}
+
+	if (buff_need > sequence->buffer_size) {
+		return false;
+	}
+
+	return true;
+}
+
+static int adc_xec_start_read(struct device *dev,
+			      const struct adc_sequence *sequence)
+{
+	struct adc_xec_regs *adc_regs = ADC_XEC_REG_BASE(dev);
+	struct adc_xec_data *data = dev->driver_data;
+
+	if (sequence->channels & ~XEC_ADC_CHANNELS_MASK) {
+		return -EINVAL;
+	}
+
+	if (sequence->channels == 0UL) {
+		return -EINVAL;
+	}
+
+	if (!adc_xec_validate_buffer_size(sequence)) {
+		return -ENOMEM;
+	}
+
+	if (sequence->resolution == 12) {
+		adc_regs->sar_control_reg.sel_res = XEC_ADC_12_BITS;
+	} else if (sequence->resolution == 10) {
+		adc_regs->sar_control_reg.sel_res = XEC_ADC_10_BITS;
+	} else {
+		return -EINVAL;
+	}
+
+	data->channels = sequence->channels;
+	data->buffer = sequence->buffer;
+	data->repeat_buffer = sequence->buffer;
```

Comment:
This will be done in [adc_context_start_sampling](https://github.com/zephyrproject-rtos/zephyr/pull/19125/commits/2f8a9dcc1ede9160c6608d26b6da0ef896dccdcf#diff-b669b6c698c20ba944cfe3b31f01dea9R106) before every sampling, so it is redundant to do it here.

Reply:
Removed.